rfd/0008-application-access.md

```diff
@@ -0,0 +1,109 @@
+This document contains technical implementation details of Teleport Application Access.
+
+## Why
+
+## Use Cases
+
+The initial implementation of Teleport Application Access is targeted at users that would like to expose internal applications and dashboards on the public internet.
+
+## Details
+
+### Identity Headers
+
+As described in a previous section, Teleport uses TLS mutual authentication to pass identity information between internal components. However, identity information is passed to proxied applications in the form of a signed JWT in a request header named `teleport-jwt-assertion`.
+
+This identity information can be used to show the identity of the user currently logged in as well as change the state of the internal application. For example, because Teleport roles are forwarded to proxied applications within the JWT header, an control panel application could show an regular or admin view based on the Teleport identity of the user.
+
+#### Issuance
+
+All Teleport clusters have a User and Host CA used to issue user and host SSH and TLS certificates. Teleport Application Access introduces a JWT signer to each cluster to issue JWTs. The JWT signer uses 2048-bit RSA keys similar to the existing CAs.
+
+#### Verification
+
+An unauthenticated endpoint at `https://proxy.example.com:3080/.well-known/jwks.json` returns the public keys that can be used to verify the signed JWT. Multiple keys are supported because JWT signers can be rotated similar to CAs.
+
+The keys are in the [JWK](https://tools.ietf.org/html/rfc7517) format. Example output of this endpoint is below.
+
+```
+$ curl https://proxy.example.com:3080/.well-known/jwks.json | jq .
+{
+  "keys": [
+    {
+      "kty": "rsa",
+      "alg": "RS256",
+      "n": "xk-0VSVZY76QGqeN9TD-FJp32s8jZrpsalnRoFwlZ_JwPbbd5-_bPKcz8o2tv1eJS0Ll6ePxRCyK68Jz2UC4V4RiYaqJCRq_qVpDQMB1sQ7p9M-8qvT82FJ-Rv-W4RNe3xRmBSFDYdXaFm51Uk8OIYfv-oZ0kGptKpkNY390aJOzjHPH2MqSvhk9Xn8GwM8kEbpSllavdJCRPCeNVGJXiSCsWrOA_wsv_jqBP6g3UOA9GnI8R6HR14OxV3C184vb3NxIqxtrW0C4W6UtSbMDcKcNCgajq2l56pHO8In5GoPCrHqlo379LE5QqpXeeHj8uqcjeGdxXTuPrRq1AuBpvQ",
+      "e": "AQAB"
+    }
+  ]
+}
+```
+
+While the RFC explains the format of these keys and how parse and verify a JWT token, we strongly recommend you use a well written and supported library in the language of your choice to validate the JWT and you not try to write parsing and validation code yourself.
+
+We have provided an example within Teleport on how to [validate the JWT token](https://github.com/gravitational/teleport/blob/a7fca6bf2bc15fc487620b410151c2108ac1d6cc/lib/jwt/jwt.go#L192-L225) written in Go.
+
+#### Claims
+
+The JWT embeds within it claims about the identity of the subject and issuer of the token.
+
+The following public claims are included:
+
+* `aud`: Audience of JWT. This is the URI of the proxied application to which the request is being forwarded.
+* `exp`: Expiration time of the JWT. This value is always in sync with the expiration of the TLS certificate.
```

Comment:
Is this the same as a `max_session_ttl` set in the role?

Reply:
Yes, it is derived from that.